lib/active_model/serializer/adapter/json_api.rb

```diff
@@ -44,21 +44,36 @@ def object
         def initialize(serializer, options = {})
           super
           @include_tree = IncludeTree.from_include_args(options[:include])
-          @fieldset = options[:fieldset] || ActiveModel::Serializer::Fieldset.new(options.delete(:fields))
+          @fieldset = options[:fieldset] || ActiveModel::Serializer::Fieldset.new(options[:fields])
         end
 
+        # Build JSON API document.
         def serializable_hash(options = nil)
           options ||= {}
 
-          hash =
-            if serializer.respond_to?(:each)
-              serializable_hash_for_collection(options)
+          resources = resource_objects_for(serializer)
+          primary_data, included = resources.each_with_object([[], []]) do |(_, resource), (prim, incl)|
+            is_primary = resource.delete(:is_primary)
+            if is_primary
+              prim.push(resource)
             else
-              serializable_hash_for_single_resource
+              incl.push(resource)
             end
+          end
+
+          is_collection = serializer.respond_to?(:each)
+          hash = {}
+          hash[:data] = is_collection ? primary_data : primary_data.first
```

Comment:
so, can you write some inline comments about that then?
about what data is and how it's intended to only be an array and how the state of primary data plays into that intention?